rfc/src/rfcs/0004-loop-contract-synthesis.md

```diff
@@ -0,0 +1,209 @@
+
+
+Also, loop contracts could improve Kani’s verification time since all loops will be abstracted to a single iteration, as opposed to being unwound a large number of iterations.
+For example, we can easily find out that the following loop is bounded by an unwinding value of `5000`.
+Kani can verify the program in a few minutes by unwinding the loop 5000 times.
+With loop contracts, we only need to verify the single abstract iteration of the loop, which leads to a smaller query.
+As a result, Kani with the synthesizer can verify the program in a few seconds.
+```rust
+#[kani::proof]
+#[kani::unwind(5000)]
+fn main() {
+    let mut y: i32 = 5000;
+
+    while y > 0 {
+        y = y - 1;
+    }
+    assert!(y == 0);
+}
+```
+
+The `goto-synthesizer` is an [enumeration-based synthesizer](https://www.cis.upenn.edu/~alur/SyGuS13.pdf).
+It enumerates candidate invariants from a pre-designed search space described by a given regular tree grammar and verifies if the candidate is an inductive invariant.
+Therefore it has the following limitations:
+1. the search space is not complete, so it may fail to find a working candidate. The current search space consists of only conjunction of linear inequalities built from the variables in the loop, which is not expressive enough to capture all loop invariants.
+For example, the loop invariant `a[i] == 0` contains array access and cannot be captured by the current search space.
+However, we can easily extend the search space to include more complex expressions with the cost of an exponential increase of the running time of the synthesizer.
+2. the synthesizer suffers from the same limitation as the loop contract verification in CBMC. For example, it does not support unbounded quantifiers, or dynamic allocations in the loop body. 
+
+## User Experience
+
+Once this RFC has been stabilized, users will be able to use the new command-line flag `--synthesize-loop-contracts` to run the synthesizer, which will attempt to synthesize loop contracts, and verify programs with the synthesized loop contracts.
+
+
+#### Limit Resource Used by Synthesizer for Termination
+Without a resource limit, an enumerative synthesizer may run forever to exhaust a search space consisting of an infinite number of candidates, especially when there is no solution in the search space.
+So, for the guarantee of termination, we provide users options: `--limit-synthesis-time T` to limit the running time of the synthesizer.
+
+
+#### Output of Kani when the Synthesizer is Enabled
+When the flag `--synthesize-loop-contracts` is provided, Kani will report different result for different cases
+1. When there exists some loop invariant in the candidate space with which all assertions can be proved, Kani will synthesize the loop contracts, verify the program with the synthesized loop contracts, and report verification SUCCESS;
```

Comment:
Question: Are programs with failures (e.g., assertions) not good candidates for loop-contract synthesis? In other words, is it possible to deduce loop contracts for programs which would result in `FAILURE`?

Reply:
Loop contracts over-approximate the states that we can reach at any iteration of the loops. 
We say a candidate is valid if it is inductive and it satisfy the base case.
But if there are still violated property, we can't distinguish whether the candidate is not strong enough to prove the property, or the property is always violated with no matter what loop contracts.
So we say a candidate good only when there is no violation.

Synthesis with limited resource will return the strongest valid candidate found when the limit reached.